### DISCUSSION TITLE
Extra indentation at function definition

### DISCUSSION THREAD

**Discussion by jsh9** [jsh9/cercis#15] [General]
Hi @jni, @flexatone, @Martmists-GH, @stefanoborini, @mintchkin, @nbievre:

I recently forked [Black](https://github.com/psf/black) and wrote a more configurable Python code formatter (named "**Cercis**": https://github.com/jsh9/cercis), because I do not agree with this particular style choice of Black, and Black does not offer customization for it:

```python
def some_function(
    arg1: int,
    arg2: bool,
    arg3: float,
    arg4_with_long_name: str,
) -> bool:
    ...
```

After some Googling, I think you share the same style opinion as I do.  Here are the posts that I looked at:

* https://github.com/grantjenks/blue/issues/89
* https://github.com/grantjenks/blue/issues/90
* https://github.com/psf/black/issues/1127
* https://github.com/psf/black/issues/2117
* https://github.com/psf/black/issues/1178

----

Cercis does this out of the box:

```python
def some_function(
        arg1: int,
        arg2: bool,
        arg3: float,
        arg4_with_long_name: str,
) -> bool:
    ...
```

It is controlled by an option [`--function-definition-extra-indent`](https://github.com/jsh9/cercis#33-extra-indentation-at-function-definition), which is by default `True`.

With the [`--closing-bracket-indentation`](https://github.com/jsh9/cercis#34-closing-bracket-indentation) option set to `True` (by default, this option is `False`), we can get this:

```python
def some_function(
        arg1: int,
        arg2: bool,
        arg3: float,
        arg4_with_long_name: str,
        ) -> bool:
    ...
```

If you are still interested in implementing this style in your code, please feel free to try it out, and any feedback would be appreciated!

**Comment by jni** [comment]
Brilliant! Thanks for the ping! I'll give it a try. Many of the projects I work with are much bigger so I can't control a change away from black, but I'll give this a go with new projects! 🙏

**Comment by jni** [comment]
@jsh9 does this also do extra indentation for line *continuation* in general? ie I like to use 8 spaces not just for function def but function calls, multi-line conditionals, etc.

**Reply by jni** [reply]
And the other thing that really bugs me about black is that it splits lines on expression parentheses rather than on conditional operators:

```python
if predicate1(
    value
) and predicate2(value):
```

👆 black's preference 🤢 vs:

```python
if (predicate1(value)
        and predicate2(value)):
```

or


```python
if (
        predicate1(value)
        and predicate2(value)
        ):
```

**Reply by jsh9** [reply]
Hi @jni:

> I like to use 8 spaces not just for function def but function calls, multi-line conditionals, etc.

It's definitely possible to implement this.  The 4 space indentation is defined [right here](https://github.com/jsh9/cercis/blob/8ce1cceb7d819a529a775d9dee11754eb2983c8d/src/cercis/lines.py#L507).

I can make it 8 spaces.  I suppose if we use 8 spaces, you don't need **_extra_** indentation (16 spaces) at function definition?

That said, Python generally has a rather small line length limit (79, or 88 in Black).  Using 8 spaces will produce more line wrapping.  So I'm curious what your rationale of using 8 spaces is?

Btw, Java (at least IntelliJ) uses 8-space indents.  That's generally fine because the conventional line length limit for Java is 120 characters.

> And the other thing that really bugs me about black is that it splits lines on expression parentheses rather than on conditional operators

I agree that the default style of Black here is not great.  It may not be very straightforward to implement though.  Let me look into this to see how difficult it can be.

**Reply by jsh9** [reply]
If I'm to implement 8 spaces, I might as well also support options of 2 spaces and tabs.

For the option of 2 spaces, what does everything think of the following style?

(There are 2 spaces in the argument list, rather than 4, or 8)
```python
def function(
  arg1_with_very_long_name: int,
  arg2_with_very_long_name: float,
  arg3_with_very_long_name: list,
):
  print('Yes')
```
(There are 2 spaces in the argument list, rather than 4, or 8.)

I'm personally OK with this style, because the 3 input arguments have a visually clean separation from `def function`.

**Reply by jni** [reply]
No no, my point is not to make all indentation 8 spaces, but to make all *line continuation* 8 spaces, to distinguish it from logical blocks. This is what PEP8 recommends.

**Reply by jsh9** [reply]
I see.  I couldn't find the recommendation of using 8 spaces in line continuation in PEP8 (https://peps.python.org/pep-0008/).  Could you point me to it?

This style you are suggesting (8 spaces for line continuation and 4 spaces for blocks) is indeed the default style for Java in IntelliJ, so I can definitely see the appeal of it.  Let me look into how to implement it.

**Reply by jsh9** [reply]
Hi @jni, please check out version 0.1.5.  I implemented configurable indentation for line continuation.

There are now 3 options for granular tweaking the code style:

* `--function-definition-extra-indent`
* `--other-line-continuation-extra-indent`
* `--closing-bracket-extra-indent`

The 2nd one is what you requested.

I also added a `--base-indentation-spaces` option.

Here is a quick example:

```python
# --base-indentation-spaces = 2
# --function-definition-extra-indent = False
# --other-line-continuation-extra-indent = True
# --closing-bracket-extra-indent = True
def some_func(
  arg1,
  arg2,
  ):
  print(1)

result = [
    1,
    2,
    ]
```

Another example:

```python
# --base-indentation-spaces = 3
# --function-definition-extra-indent = True
# --other-line-continuation-extra-indent = False
# --closing-bracket-extra-indent = True
def some_func(
      arg1,  # indent 6 spaces here
      arg2,
      ):
   print(1)

result = [
   1,  # indent 3 spaces here
   2,
   ]
```

**Reply by jni** [reply]
Thanks! Indeed I see looking more closely at the PEP8 examples that I had overgeneralised from this statement:

>  When using a hanging indent the following should be considered; there should be no arguments on the first line and **further indentation should be used to clearly distinguish itself as a continuation line**:

But this is what I'm talking about:

```python
# Add some extra indentation on the conditional continuation line.
if (this_is_one_thing
        and that_is_another_thing):
    do_something()
```

Basically the argument here:

```python
# Add 4 spaces (an extra level of indentation) to distinguish arguments from the rest.
def long_function_name(
        var_one, var_two, var_three,
        var_four):
    print(var_one)
```

Applies to any block that is indented, including with, while, for, if;

I find the 8 spaces indents for continuation (any continuation) useful to help me visually group lines — 4 spaces -> code block; 8 spaces -> mentally note this is a long line. I can glance at a function and count the number of statements, which is harder to do if line continuations are four spaces like the function def.

Thanks for the work!!! ❤️

**Comment by stefanoborini** [comment]
about damn time that someone resists that formatting cult that is black. Well done.

**Comment by jsh9** [comment]
Hi @veritas9872, @nilsbunger:

I saw your posts (https://github.com/astral-sh/ruff/issues/1904#issuecomment-1515784435) and/or upvote to request extra indentation in the function argument list.

In particular, @veritas9872 mentioned [here](https://github.com/astral-sh/ruff/issues/1904#issuecomment-1517133555) that:

> Black continues to use its current indentation for arguments as changing it would be too disruptive

In fact, it is not difficult to implement, and it is not disruptive.  I forked `Black` a few weeks ago and added this configurable option.

Please feel free to check out https://github.com/jsh9/cercis, and the option I added is : [--function-definition-extra-indent](https://github.com/jsh9/cercis#351-at-function-definition---function-definition-extra-indent).

My fork is on top of the latest changes of `Black` and I plan to keep rebasing on top of it.

Please feel free to take a look and provide feedback if any!

**Comment by jsh9** [comment]
Hi @jni, @flexatone, @Martmists-GH, @stefanoborini, @mintchkin, @nbievre, @veritas9872, @nilsbunger:

Ruff is implementing a Python code formatter, and I think the authors of Ruff are open-minded about adding configuration options.  If you'd like this 8-space indentation feature implemented in Ruff, would you mind going to upvote my comment here:: https://github.com/astral-sh/ruff/issues/1904#issuecomment-1716294595

Thanks!

**Reply by jsh9** [reply]
I think using Ruff is a better long-term solution than forking `Black` and making changes.

**Reply by jni** [reply]
I agree but by my reading the Ruff developers' willingness to implement options is quite limited. Anyway, I've upvoted. 🙏 Thank you again for your work! 🙏

**Reply by jsh9** [reply]
This time it may not be that easy for me to fork and tweak, because I don't code in Rust at all. :-(

**Reply by jni** [reply]
Sure, but it's not like black is going to disappear. I'd rather use cercis than black-from-ruff. 😜 So I hope you still have your original plan...

**Comment by jsh9** [comment]
Hi @leonardder, welcome to the club!  Please take a look at this discussion thread and the repo.
